Unable to update records in uncompressed chunk of hypertable with prepared statements

**Relevant system information:**
 - OS: Ubuntu 18.04, CentOS Linux release 7.9.2009 (Core)
 - PostgreSQL version (output of `postgres --version`): 12.8
 - TimescaleDB version (output of `\dx` in `psql`): 2.4.1, 2.3.1, 2.0.1
 - Installation method: apt install, AWS AMI image

**Describe the bug**
When I update records in an uncompressed chunk of a hypertable, I get an error saying update failed because I cannot update/delete rows from a chunk that has been compressed. However, the record that I am trying to update is not in the chunk referenced in the error message. 

**To Reproduce**
I have a simple Java program to reproduce the problem. The program uses JDBC and executes all inserts and updates with the same connection. I will describe the sequence of operations below. I have also attached the test program.

1. Create a hypertable with this definition:
```
create table test_ht (
  id int,
  time timestamptz NOT NULL,
  x varchar(10),
  PRIMARY KEY (id, time)
);
SELECT create_hypertable('test_ht', 'time', chunk_time_interval => INTERVAL '1 month');
ALTER TABLE test_ht SET (                                                                 
  timescaledb.compress,
  timescaledb.compress_segmentby = 'id'
);
```
2. Insert 1 row into the table with timestamp = '2021-01-01 00:00:00+00:00'. 
```
insert into test_ht (id, time, x) values (1, '2021-01-01 00:00:00+00:00', 'a');
```
3. Compress the chunk that contains the record added. The parameter to the compress_chunk function is the schema and name of the chunk, which is different for each test.  
```
SELECT public.compress_chunk('_timescaledb_internal._hyper_9_8_chunk');
```
4. Insert 1000 records, all with timestamp '2021-02-22T08:00:00+00:00'. I do that in a loop. 
```
insert into test_ht (id, time, x) values (2, '2021-02-22 08:00:00+00:00', 'a');
insert into test_ht (id, time, x) values (3, '2021-02-22 08:00:00+00:00', 'a');
...
...
insert into test_ht (id, time, x) values (1000, '2021-02-22 08:00:00+00:00', 'a');
insert into test_ht (id, time, x) values (1001, '2021-02-22 08:00:00+00:00', 'a');
```
4. Update each of the records that has timestamp '2021-02-22 08:00:00+00:00' by changing the value of column `x` to a different character. Each row is updated one at a time. 
```
update test_ht set x = 'x' where time='2021-02-22T08:00:00+00:00' and id = 2;
update test_ht set x = 'x' where time='2021-02-22T08:00:00+00:00' and id = 3;
...
...
update test_ht set x = 'x' where time='2021-02-22T08:00:00+00:00' and id = 100;
update test_ht set x = 'x' where time='2021-02-22T08:00:00+00:00' and id = 101;
```


**Expected behavior**
I expect all insert and update statements to be executed successfully. All updates are done on record with timestamps in February, which is not in the compressed chunk. 

Here are the two chunks created in the hypertable after executing the above operations:
```
scopedb=> select chunk_schema,chunk_name,range_start, range_end,is_compressed from timescaledb_information.chunks c where c.hypertable_schema = 'scopedb' AND c.hypertable_name = 'test_ht';
     chunk_schema      |     chunk_name     |      range_start       |       range_end        | is_compressed 
-----------------------+--------------------+------------------------+------------------------+---------------
_timescaledb_internal | _hyper_43_54_chunk | 2020-12-04 00:00:00+00 | 2021-01-03 00:00:00+00 | t
_timescaledb_internal | _hyper_43_56_chunk | 2021-02-02 00:00:00+00 | 2021-03-04 00:00:00+00 | f
(2 rows)
```

**Actual behavior**
Here is the error I got:
```
java.sql.BatchUpdateException: Batch entry 5 update test_ht set x = 'x' where time='2021-02-22 03:00:00-05' and id = 7 was aborted: ERROR: cannot update/delete rows from chunk "_hyper_43_54_chunk" as it is compressed  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleCompletion(BatchResultHandler.java:199)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:531)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:878)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:901)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1644)
	at com.vmturbo.scope.replay.TestHT.loadData(TestHT.java:106)
	at com.vmturbo.scope.replay.TestHT.main(TestHT.java:123)
Caused by: org.postgresql.util.PSQLException: ERROR: cannot update/delete rows from chunk "_hyper_43_54_chunk" as it is compressed
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2552)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2284)
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1403)
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1428)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:506)
	... 5 more
```
The chunk referenced in the error message is the one that was compressed, but it contains only 1 record that has a timestamp in January. The record I am trying to upgrade has timestamp in February, which is in the chunk `_hyper_43_56_chunk`.

I have tried the test described above with varying number of records in the 2 chunks. The tests fails with a relatively smaller number of rows in the compressed chunk compared to the uncompressed chunk.

If I have 1000 records in the uncompressed chunk, the test fails with up to 566 records in the first chunk. If I have 567 or more rows in the first (compressed) chunk, the test program can complete without errors. 

If I have 10,000 records in the uncompressed chunk, the test fails when there are less than or equal 566 records in the first chunk.

If I have 100,000 records in the uncompressed chunk, the test fails when there are less than or equal 574 records in the first chunk.

For step #4 of the steps to reproduce, if I update all records in one update statement, there are no errors. 
e.g. 
```
update test_ht set x = 'x' where time='2021-02-02 00:00:00+00'
```

Test Program:
[test.zip](https://github.com/timescale/timescaledb/files/7375768/test.zip)

**Screenshots**
N/A

**Additional context**
N/A

--- a/src/nodes/chunk_append/exec.c
+++ b/src/nodes/chunk_append/exec.c
@@ -176,7 +176,7 @@ do_startup_exclusion(ChunkAppendState *state)
 	 * create skeleton plannerinfo for estimate_expression_value
 	 */
 	PlannerGlobal glob = {
-		.boundParams = NULL,
+		.boundParams = state->csstate.ss.ps.state->es_param_list_info,
 	};
 	PlannerInfo root = {
 		.glob = &glob,
@@ -339,7 +339,7 @@ initialize_runtime_exclusion(ChunkAppendState *state)
 	int i = 0;
 
 	PlannerGlobal glob = {
-		.boundParams = NULL,
+		.boundParams = state->csstate.ss.ps.state->es_param_list_info,
 	};
 	PlannerInfo root = {
 		.glob = &glob,

--- a/tsl/test/shared/expected/compression_dml.out
+++ b/tsl/test/shared/expected/compression_dml.out
@@ -1,0 +1,26 @@
+-- This file and its contents are licensed under the Timescale License.
+-- Please see the included NOTICE for copyright information and
+-- LICENSE-TIMESCALE for a copy of the license.
+-- test constraint exclusion with prepared statements and generic plans
+CREATE TABLE i3719 (time timestamptz NOT NULL,data text);
+SELECT table_name FROM create_hypertable('i3719', 'time');
+ table_name 
+ i3719
+(1 row)
+
+ALTER TABLE i3719 SET (timescaledb.compress);
+INSERT INTO i3719 VALUES('2021-01-01 00:00:00', 'chunk 1');
+SELECT count(compress_chunk(c)) FROM show_chunks('i3719') c;
+ count 
+     1
+(1 row)
+
+INSERT INTO i3719 VALUES('2021-02-22 08:00:00', 'chunk 2');
+SET plan_cache_mode TO force_generic_plan;
+PREPARE p1(timestamptz) AS UPDATE i3719 SET data = 'x' WHERE time=$1;
+PREPARE p2(timestamptz) AS DELETE FROM i3719 WHERE time=$1;
+EXECUTE p1('2021-02-22T08:00:00+00');
+EXECUTE p2('2021-02-22T08:00:00+00');
+DEALLOCATE p1;
+DEALLOCATE p2;
+DROP TABLE i3719;

--- a/tsl/test/shared/expected/constraint_exclusion_prepared.out
+++ b/tsl/test/shared/expected/constraint_exclusion_prepared.out
@@ -541,6 +541,47 @@ QUERY PLAN
 
 DEALLOCATE prep;
 RESET timescaledb.enable_chunk_append;
+-- test prepared statement with params and generic plan
+SET plan_cache_mode TO force_generic_plan;
+PREPARE prep(timestamptz) AS SELECT device_id, time FROM :TEST_TABLE WHERE time = $1;
+:PREFIX EXECUTE prep('2000-01-01 23:42');
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics (actual rows=5 loops=1)
+   Chunks excluded during runtime: 2
+   ->  Index Scan using _hyper_X_X_chunk_metrics_time_idx on _hyper_X_X_chunk (actual rows=5 loops=1)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+(8 rows)
+
+:PREFIX EXECUTE prep('2000-01-10 23:42');
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics (actual rows=5 loops=1)
+   Chunks excluded during runtime: 2
+   ->  Index Scan using _hyper_X_X_chunk_metrics_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_time_idx on _hyper_X_X_chunk (actual rows=5 loops=1)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+(8 rows)
+
+:PREFIX EXECUTE prep(now());
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics (actual rows=0 loops=1)
+   Chunks excluded during runtime: 3
+   ->  Index Scan using _hyper_X_X_chunk_metrics_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+(8 rows)
+
+DEALLOCATE prep;
+RESET plan_cache_mode;
 \set TEST_TABLE 'metrics_space'
 \ir :TEST_QUERY_NAME
 -- This file and its contents are licensed under the Timescale License.
@@ -1336,6 +1377,83 @@ QUERY PLAN
 
 DEALLOCATE prep;
 RESET timescaledb.enable_chunk_append;
+-- test prepared statement with params and generic plan
+SET plan_cache_mode TO force_generic_plan;
+PREPARE prep(timestamptz) AS SELECT device_id, time FROM :TEST_TABLE WHERE time = $1;
+:PREFIX EXECUTE prep('2000-01-01 23:42');
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics_space (actual rows=5 loops=1)
+   Chunks excluded during runtime: 6
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (actual rows=1 loops=1)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (actual rows=3 loops=1)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (actual rows=1 loops=1)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+(20 rows)
+
+:PREFIX EXECUTE prep('2000-01-10 23:42');
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics_space (actual rows=5 loops=1)
+   Chunks excluded during runtime: 6
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (actual rows=1 loops=1)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (actual rows=3 loops=1)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (actual rows=1 loops=1)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+(20 rows)
+
+:PREFIX EXECUTE prep(now());
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics_space (actual rows=0 loops=1)
+   Chunks excluded during runtime: 9
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+   ->  Index Scan using _hyper_X_X_chunk_metrics_space_time_idx on _hyper_X_X_chunk (never executed)
+         Index Cond: ("time" = $1)
+(20 rows)
+
+DEALLOCATE prep;
+RESET plan_cache_mode;
 \set TEST_TABLE 'metrics_compressed'
 \ir :TEST_QUERY_NAME
 -- This file and its contents are licensed under the Timescale License.
@@ -1946,6 +2064,69 @@ QUERY PLAN
 
 DEALLOCATE prep;
 RESET timescaledb.enable_chunk_append;
+-- test prepared statement with params and generic plan
+SET plan_cache_mode TO force_generic_plan;
+PREPARE prep(timestamptz) AS SELECT device_id, time FROM :TEST_TABLE WHERE time = $1;
+:PREFIX EXECUTE prep('2000-01-01 23:42');
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics_compressed (actual rows=5 loops=1)
+   Chunks excluded during runtime: 2
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (actual rows=5 loops=1)
+         Filter: ("time" = $1)
+         Rows Removed by Filter: 4995
+         ->  Seq Scan on compress_hyper_X_X_chunk (actual rows=5 loops=1)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+               Rows Removed by Filter: 15
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+(16 rows)
+
+:PREFIX EXECUTE prep('2000-01-10 23:42');
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics_compressed (actual rows=5 loops=1)
+   Chunks excluded during runtime: 2
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (actual rows=5 loops=1)
+         Filter: ("time" = $1)
+         Rows Removed by Filter: 4995
+         ->  Seq Scan on compress_hyper_X_X_chunk (actual rows=5 loops=1)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+               Rows Removed by Filter: 25
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+(16 rows)
+
+:PREFIX EXECUTE prep(now());
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics_compressed (actual rows=0 loops=1)
+   Chunks excluded during runtime: 3
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+(14 rows)
+
+DEALLOCATE prep;
+RESET plan_cache_mode;
 \set TEST_TABLE 'metrics_space_compressed'
 \ir :TEST_QUERY_NAME
 -- This file and its contents are licensed under the Timescale License.
@@ -2996,6 +3177,149 @@ QUERY PLAN
 
 DEALLOCATE prep;
 RESET timescaledb.enable_chunk_append;
+-- test prepared statement with params and generic plan
+SET plan_cache_mode TO force_generic_plan;
+PREPARE prep(timestamptz) AS SELECT device_id, time FROM :TEST_TABLE WHERE time = $1;
+:PREFIX EXECUTE prep('2000-01-01 23:42');
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics_space_compressed (actual rows=5 loops=1)
+   Chunks excluded during runtime: 6
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (actual rows=1 loops=1)
+         Filter: ("time" = $1)
+         Rows Removed by Filter: 999
+         ->  Seq Scan on compress_hyper_X_X_chunk (actual rows=1 loops=1)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+               Rows Removed by Filter: 3
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (actual rows=3 loops=1)
+         Filter: ("time" = $1)
+         Rows Removed by Filter: 2997
+         ->  Seq Scan on compress_hyper_X_X_chunk (actual rows=3 loops=1)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+               Rows Removed by Filter: 9
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (actual rows=1 loops=1)
+         Filter: ("time" = $1)
+         Rows Removed by Filter: 999
+         ->  Seq Scan on compress_hyper_X_X_chunk (actual rows=1 loops=1)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+               Rows Removed by Filter: 3
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+(44 rows)
+
+:PREFIX EXECUTE prep('2000-01-10 23:42');
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics_space_compressed (actual rows=5 loops=1)
+   Chunks excluded during runtime: 6
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (actual rows=1 loops=1)
+         Filter: ("time" = $1)
+         Rows Removed by Filter: 999
+         ->  Seq Scan on compress_hyper_X_X_chunk (actual rows=1 loops=1)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+               Rows Removed by Filter: 5
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (actual rows=3 loops=1)
+         Filter: ("time" = $1)
+         Rows Removed by Filter: 2997
+         ->  Seq Scan on compress_hyper_X_X_chunk (actual rows=3 loops=1)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+               Rows Removed by Filter: 15
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (actual rows=1 loops=1)
+         Filter: ("time" = $1)
+         Rows Removed by Filter: 999
+         ->  Seq Scan on compress_hyper_X_X_chunk (actual rows=1 loops=1)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+               Rows Removed by Filter: 5
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+(44 rows)
+
+:PREFIX EXECUTE prep(now());
+QUERY PLAN
+ Custom Scan (ChunkAppend) on metrics_space_compressed (actual rows=0 loops=1)
+   Chunks excluded during runtime: 9
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+   ->  Custom Scan (DecompressChunk) on _hyper_X_X_chunk (never executed)
+         Filter: ("time" = $1)
+         ->  Seq Scan on compress_hyper_X_X_chunk (never executed)
+               Filter: ((_ts_meta_min_1 <= $1) AND (_ts_meta_max_1 >= $1))
+(38 rows)
+
+DEALLOCATE prep;
+RESET plan_cache_mode;
 -- get results for all the queries
 -- run queries on uncompressed hypertable and store result
 \set PREFIX ''

--- a/tsl/test/shared/sql/CMakeLists.txt
+++ b/tsl/test/shared/sql/CMakeLists.txt
@@ -18,7 +18,7 @@ set(TEST_TEMPLATES_SHARED
     ordered_append_join.sql.in transparent_decompress_chunk.sql.in)
 
 if((${PG_VERSION_MAJOR} GREATER_EQUAL "14"))
-  list(APPEND TEST_FILES_SHARED memoize.sql)
+  list(APPEND TEST_FILES_SHARED compression_dml.sql memoize.sql)
 endif()
 
 if(CMAKE_BUILD_TYPE MATCHES Debug)

--- a/tsl/test/shared/sql/compression_dml.sql
+++ b/tsl/test/shared/sql/compression_dml.sql
@@ -1,0 +1,23 @@
+-- This file and its contents are licensed under the Timescale License.
+-- Please see the included NOTICE for copyright information and
+-- LICENSE-TIMESCALE for a copy of the license.
+
+-- test constraint exclusion with prepared statements and generic plans
+CREATE TABLE i3719 (time timestamptz NOT NULL,data text);
+SELECT table_name FROM create_hypertable('i3719', 'time');
+ALTER TABLE i3719 SET (timescaledb.compress);
+
+INSERT INTO i3719 VALUES('2021-01-01 00:00:00', 'chunk 1');
+SELECT count(compress_chunk(c)) FROM show_chunks('i3719') c;
+INSERT INTO i3719 VALUES('2021-02-22 08:00:00', 'chunk 2');
+
+SET plan_cache_mode TO force_generic_plan;
+PREPARE p1(timestamptz) AS UPDATE i3719 SET data = 'x' WHERE time=$1;
+PREPARE p2(timestamptz) AS DELETE FROM i3719 WHERE time=$1;
+EXECUTE p1('2021-02-22T08:00:00+00');
+EXECUTE p2('2021-02-22T08:00:00+00');
+
+DEALLOCATE p1;
+DEALLOCATE p2;
+
+DROP TABLE i3719;

--- a/tsl/test/shared/sql/include/constraint_exclusion_prepared.sql
+++ b/tsl/test/shared/sql/include/constraint_exclusion_prepared.sql
@@ -102,3 +102,14 @@ DEALLOCATE prep;
 
 RESET timescaledb.enable_chunk_append;
 
+-- test prepared statement with params and generic plan
+SET plan_cache_mode TO force_generic_plan;
+PREPARE prep(timestamptz) AS SELECT device_id, time FROM :TEST_TABLE WHERE time = $1;
+
+:PREFIX EXECUTE prep('2000-01-01 23:42');
+:PREFIX EXECUTE prep('2000-01-10 23:42');
+:PREFIX EXECUTE prep(now());
+
+DEALLOCATE prep;
+RESET plan_cache_mode;
+

